src/main/utility/mod.rs

```diff
@@ -107,26 +108,27 @@ pub trait IsSync: Sync {}
 /// validates that the `Magic` is valid before running `Drop` implementations of
 /// the other fields.
 ///
-/// The MAGIC parameter should ideally be unique for each type. Consider e.g.
-/// `python3 -c 'import random; print(random.randint(0, 2**32))'`
+/// T should be the type of the struct that contains the Magic.
 #[derive(Debug)]
-pub struct Magic<const MAGIC: u32> {
+pub struct Magic<T: 'static + ?Sized> {
     #[cfg(debug_assertions)]
-    magic: u32,
+    magic: std::any::TypeId,
+    _phantom: PhantomData<T>,
 }
```

Comment:
When you use `PhantomData<T>` rust assumes that the object owns (and will drop) a `T` and makes safety assumptions about that, but that's not the case for `Magic` which doesn't access an object of that type at all. But I don't really know what you're supposed to use here. Maybe `_phantom: PhantomData<*const T>` would be slightly better [to avoid a drop check](https://doc.rust-lang.org/nomicon/phantom-data.html#table-of-phantomdata-patterns)? But I don't think it really matters either way, I just thought it was mildly interesting :)

Reply:
Huh, interesting.

If we use `* const T`, then `Magic` becomes `!Send` and `!Sync`. We could add them back, but then we'd need to be careful to ensure doing so is valid in the future. I think implementing our `IsSend` and `IsSync` traits for `TypeId` would do it... as long as we also ensured no other fields were added without also adding such a check.

Leaving as `PhantomData<T>` for now, but added a note.